src/main/java/com/uid2/shared/attest/UidCoreClient.java

```diff
@@ -1,64 +1,67 @@
     private final Proxy proxy;
-    private final String attestationEndpoint;
-    private final String userToken;
-    private final ApplicationVersion appVersion;
+    private String userToken;
     private final String appVersionHeader;
-    private AtomicReference<String> attestationToken;
-    private final boolean enforceHttps;
-    private static boolean useSecureParameters = true;
+    private boolean enforceHttps;
     private boolean allowContentFromLocalFileSystem = false;
+    private HttpClient httpClient;
+    private AttestationTokenRetriever attestationTokenRetriever;
     private AtomicReference<Handler<Integer>> responseWatcher;
 
-    public static UidCoreClient createNoAttest(String attestationEndpoint, String userToken, ApplicationVersion appVersion, boolean enforceHttps) {
+    public static UidCoreClient createNoAttest(String attestationEndpoint, String userToken, ApplicationVersion appVersion, boolean enforceHttps) throws IOException {
         return new UidCoreClient(attestationEndpoint, userToken, appVersion, CloudUtils.defaultProxy, new NoAttestationProvider(), enforceHttps);
     }
 
     public UidCoreClient(String attestationEndpoint, String userToken, ApplicationVersion appVersion, Proxy proxy,
-                         IAttestationProvider attestationProvider, boolean enforceHttps) {
-        this.attestationEndpoint = attestationEndpoint;
+                         IAttestationProvider attestationProvider, boolean enforceHttps) throws IOException {
         this.proxy = proxy;
         this.userToken = userToken;
-        this.appVersion = appVersion;
-        this.attestationProvider = attestationProvider;
         this.contentStorage = new PreSignedURLStorage(proxy);
-        this.attestationToken = new AtomicReference<>(null);
         this.enforceHttps = enforceHttps;
         this.responseWatcher = new AtomicReference<Handler<Integer>>(null);
+        this.attestationTokenRetriever = new AttestationTokenRetriever(
+                attestationEndpoint, appVersion, proxy, attestationProvider, responseWatcher, new InstantClock());
+        this.httpClient = HttpClient.newHttpClient();
+
+        String appVersionHeader = appVersion.getAppName() + "=" + appVersion.getAppVersion();
+        for (Map.Entry<String, String> kv : appVersion.getComponentVersions().entrySet())
+            appVersionHeader += ";" + kv.getKey() + "=" + kv.getValue();
+        this.appVersionHeader = appVersionHeader;
+    }
+
+    public UidCoreClient(String attestationEndpoint, String userToken, ApplicationVersion appVersion, Proxy proxy,
+                         IAttestationProvider attestationProvider, boolean enforceHttps, HttpClient httpClient) throws IOException {
+        this.proxy = proxy;
+        this.userToken = userToken;
+        this.contentStorage = new PreSignedURLStorage(proxy);
+        this.enforceHttps = enforceHttps;
+        this.responseWatcher = new AtomicReference<Handler<Integer>>(null);
```

Comment:
Why an atomic reference? Is this going to be mutated by multiple threads?

Reply:
Yes, in operator:
```
String clientsMdPath = this.config.getString(Const.Config.ClientsMetadataPathProp);
        this.clientKeyProvider = new RotatingClientKeyProvider(fsStores, new GlobalScope(new CloudPath(clientsMdPath)));
        String keysMdPath = this.config.getString(Const.Config.KeysMetadataPathProp);
        this.keyStore = new RotatingKeyStore(fsStores, new GlobalScope(new CloudPath(keysMdPath)));
        String keysAclMdPath = this.config.getString(Const.Config.KeysAclMetadataPathProp);
        this.keyAclProvider = new RotatingKeyAclProvider(fsStores, new GlobalScope(new CloudPath(keysAclMdPath)));
        String saltsMdPath = this.config.getString(Const.Config.SaltsMetadataPathProp);
        this.saltProvider = new RotatingSaltProvider(fsStores, saltsMdPath);
   ```
   It was called by 4 providers

Reply:
After talking with Thomas, I've removed `AtomicReference` for `responeWatcher` and handle the thread safety when actually calling `notifyResponseStatusWatcher`.